File: bestia-zoneserver/src/test/kotlin/net/bestia/zoneserver/map/MapServiceTest.java
package net.bestia.zoneserver.map;

import net.bestia.model.map.MapDataRepository;
import net.bestia.model.map.MapParameterRepository;
import net.bestia.model.map.MapParameter;
import net.bestia.model.geometry.Point;
import net.bestia.model.geometry.Rect;
import net.bestia.model.map.BestiaMap;
import net.bestia.model.map.MapChunk;
import net.bestia.model.map.TilesetService;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class MapServiceTest {

	private static final String MAP_NAME = "kalarian";

	private MapService ms;

	@Mock
	private MapDataRepository dataNoMapDao;

	@Mock
	private MapDataRepository dataMapDao;

	@Mock
	private MapParameterRepository paramDao;

	@Mock
	private MapParameterRepository paramNoMapDao;

	@Mock
	private MapParameter mapParams;

	@Mock
	private TilesetService tilesetService;

	@Before
	public void setup() {

		when(dataMapDao.count()).thenReturn(1L);
		when(mapParams.getName()).thenReturn(MAP_NAME);
		when(paramDao.findFirstByOrderByIdDesc()).thenReturn(mapParams);

		ms = new MapService(dataNoMapDao, paramDao, tilesetService);
	}

	@Test
	public void isMapInitialized_noMapInsideDB_false() {
		Assert.assertFalse(ms.isMapInitialized());
	}

	@Test
	public void isMapInitialized_mapInsideDB_true() {
		ms = new MapService(dataMapDao, paramDao, tilesetService);
		Assert.assertTrue(ms.isMapInitialized());
	}

	@Test(expected = IllegalArgumentException.class)
	public void getMap_illegalCoordinates_throws() {
		ms.getMap(10, 10, -10, 10);
	}

	@Test
	public void getMap_legalCoordinates_validMap() {

		BestiaMap m = ms.getMap(5, 10, 10, 10);

		Assert.assertNotNull(m);
		Assert.assertEquals(new Rect(5, 10, 10, 10), m.getRect());
	}

	@Test(expected = NullPointerException.class)
	public void saveMapData_null_throws() throws IOException {
		ms.saveMapData(null);
	}

	@Test
	public void getMapName_noMapInsideDB_emptyStr() {
		ms = new MapService(dataNoMapDao, paramNoMapDao, tilesetService);
		Assert.assertEquals("", ms.getMapName());
	}

	@Test
	public void getMapName_mapInsideDB_validStr() {
		Assert.assertEquals(MAP_NAME, ms.getMapName());
	}

	@Test(expected = NullPointerException.class)
	public void getChunks_null_throws() {
		ms.getChunks(null);
	}

	@Test
	public void getChunks_validCords_listWithChunks() {
		List<Point> chunkCords = new ArrayList<>();
		chunkCords.add(new Point(1, 1));
		List<MapChunk> chunks = ms.getChunks(chunkCords);

		Assert.assertNotNull(chunks);
		Assert.assertEquals(1, chunks.size());
	}
}

File: bestia-zoneserver/src/test/kotlin/net/bestia/zoneserver/map/MapServiceTest.kt
package net.bestia.zoneserver.map

import com.nhaarman.mockitokotlin2.whenever
import net.bestia.model.geometry.Point
import net.bestia.model.geometry.Rect
import net.bestia.model.map.MapDataRepository
import net.bestia.model.map.MapParameter
import net.bestia.model.map.MapParameterRepository
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import java.lang.IllegalArgumentException
import java.util.*

@ExtendWith(MockitoExtension::class)
class MapServiceTest {

  private lateinit var ms: MapService

  @Mock
  private lateinit var dataNoMapDao: MapDataRepository

  @Mock
  private lateinit var dataMapDao: MapDataRepository

  @Mock
  private lateinit var paramDao: MapParameterRepository

  @Mock
  private lateinit var paramNoMapDao: MapParameterRepository

  @Mock
  private lateinit var mapParams: MapParameter

  @Mock
  private lateinit var tilesetService: TilesetService

  @BeforeEach
  fun setup() {
        ms = MapService(dataNoMapDao, paramDao, tilesetService)
  }

  @Test
  fun isMapInitialized_noMapInsideDB_false() {
    Assertions.assertFalse(ms.isMapInitialized)
  }

  @Test
  fun isMapInitialized_mapInsideDB_true() {
    whenever(dataMapDao.count()).thenReturn(1L)

    ms = MapService(dataMapDao, paramDao, tilesetService)
    Assertions.assertTrue(ms.isMapInitialized)
  }

  @Test
  fun getMap_illegalCoordinates_throws() {
    Assertions.assertThrows(IllegalArgumentException::class.java) {
      ms.getMap(10, 10, -10, 10)
    }
  }

  @Disabled("We need a valid, serialized MapData here")
  @Test
  fun getMap_legalCoordinates_validMap() {
    whenever(dataMapDao.count()).thenReturn(1L)
    whenever(mapParams.name).thenReturn(MAP_NAME)
    whenever(paramDao.findFirstByOrderByIdDesc()).thenReturn(mapParams)

    val m = ms.getMap(5, 10, 10, 10)

    Assertions.assertNotNull(m)
    Assertions.assertEquals(Rect(5, 10, 10, 10), m.rect)
  }

  @Test
  fun getMapName_noMapInsideDB_emptyStr() {
    ms = MapService(dataNoMapDao, paramNoMapDao, tilesetService)
    Assertions.assertEquals("", ms.mapName)
  }

  @Test
  fun getMapName_mapInsideDB_validStr() {
    whenever(mapParams.name).thenReturn(MAP_NAME)
    whenever(paramDao.findFirstByOrderByIdDesc()).thenReturn(mapParams)

    Assertions.assertEquals(MAP_NAME, ms.mapName)
  }

  @Test
  fun getChunks_validCords_listWithChunks() {
    val chunkCords = ArrayList<Point>()
    chunkCords.add(Point(1, 1))
    val chunks = ms.getChunks(chunkCords)

    Assertions.assertNotNull(chunks)
    Assertions.assertEquals(1, chunks.size.toLong())
  }

  companion object {
    private const val MAP_NAME = "kalarian"
  }
}
